include/sundials/sundials_types.h

```diff
@@ -178,19 +178,59 @@ typedef SUNDIALS_INDEX_TYPE sunindextype;
 
 /*
  *------------------------------------------------------------------
- * Type : sunoutputformat
+ * Type : SUNOutputFormat
  *------------------------------------------------------------------
  * Constants for different output formats
  *------------------------------------------------------------------
  */
 
-typedef enum {
+typedef enum
+{
   SUN_OUTPUTFORMAT_TABLE,
   SUN_OUTPUTFORMAT_CSV
 } SUNOutputFormat;
 
+/*
+ *------------------------------------------------------------------
+ * Type : SUNErrCode
+ *------------------------------------------------------------------
+ * Error code type
+ *------------------------------------------------------------------
+ */
+
+typedef int SUNErrCode;
+
+/*
+ *------------------------------------------------------------------
+ * Type : SUNNlsStatus
+ *------------------------------------------------------------------
+ * Status code returned by some SUNNonLinearSolver functions.
+ * Different than SUNErrCode, as this is used to indicate numerical
+ * errors that may be recoverable or unrecoverable (e.g. convergence).
+ *------------------------------------------------------------------
+ */
+
+typedef int SUNNlsStatus;
+
+/*
+ *------------------------------------------------------------------
+ * Type : SUNLsStatus
+ *------------------------------------------------------------------
+ * Status code returned by some SUNLinearSolver functions.
+ * Different than SUNErrCode, as this is used to indicate numerical
+ * errors that may be recoverable or unrecoverable (e.g. convergence).
+ *------------------------------------------------------------------
+ */
+
+/* TODO(CJB): some functions return an index as a SUNLsStatus due to
+   backwards compatibility. This could result in overflow if the
+   index returned is larger than an int. We could make SUNLsStatus
+   a sunindextype, but then updating functions like PSetup and PSolve would
+   technically be API changes. */
```

Comment:
I think that this is a vestige of the previous non-object-oriented implementations, wherein all data was pre-declared and all arguments were passed in the function call, so errors could not really be "stored" so much as returned.  This, combined with the fact that when they were written, they emulated 32-bit integer LAPACK functionality.

I think that the most elegant fix would be to have our direct SUNLinearSolver modules store the relevant index internally, and return a relevant `SUNErrCode` to indicate a SUNLinearSolver failure.  Users could then directly query the SUNLinearSolver module to find out what happened.

Reply:
I agree. We can do this later with a major version bump.